pallets/phala/src/stakepool.rs

```diff
@@ -471,6 +559,95 @@ pub mod pallet {
+					.saturated_into::<u64>();
+				// all of the free_stake would be withdrew back to user
+				let unwithdraw_amount = amount.saturating_sub(pool_info.free_stake);
+				pool_info.free_stake = Zero::zero();
+				pool_info.total_stake = pool_info.total_stake.saturating_sub(pool_info.free_stake);
+				user_info.amount = user_info.amount.saturating_sub(unwithdraw_amount);
+
+				// case some locked asset has not been withdraw(unlock) to user, add it to withdraw queue.
+				// when pool has free stake again, the withdraw would be handled
+				pool_info.withdraw_queue.push_back(WithdrawInfo {
+					user: user_info.user.clone(),
+					amount: unwithdraw_amount,
+					start_time: now,
+				});
+			}
+			user_info.user_debt =
+				user_info.amount * pool_info.pool_acc / 10u32.pow(6).saturated_into();
+
+			// update the lock balance of user
+			Self::update_lock(user_info.user.clone(), user_info.amount);
+		}
+
+		fn try_handle_waitting_withdraw(pool_info: &mut PoolInfo<T::AccountId, BalanceOf<T>>) {
+			while pool_info.free_stake > Zero::zero() {
+				if pool_info.withdraw_queue.is_empty() {
+					break;
+				}
+
+				if let Some(mut withdraw_info) = pool_info.withdraw_queue.pop_front() {
+					let info_key = (pool_info.pid.clone(), withdraw_info.user.clone());
+					let mut user_info = Self::staking_info(&info_key).unwrap();
+
+					if pool_info.free_stake <= withdraw_info.amount {
+						pool_info.free_stake = Zero::zero();
+						pool_info.total_stake =
+							pool_info.total_stake.saturating_sub(pool_info.free_stake);
+						withdraw_info.amount =
+							withdraw_info.amount.saturating_sub(pool_info.free_stake);
+
+						// push front the updated withdraw info
+						pool_info.withdraw_queue.push_front(withdraw_info);
+
+						// update user_debt which would determine the user's rewards
+						user_info.amount = user_info.amount.saturating_sub(pool_info.free_stake);
+						user_info.user_debt =
+							user_info.amount * pool_info.pool_acc / 10u32.pow(6).saturated_into();
+					} else {
+						// all of the amount would be withdraw to user and no need to push the popped one back
+						pool_info.free_stake =
+							pool_info.free_stake.saturating_sub(withdraw_info.amount);
+						pool_info.total_stake =
+							pool_info.total_stake.saturating_sub(withdraw_info.amount);
+
+						// update user_debt which would determine the user's rewards
+						user_info.amount = user_info.amount.saturating_sub(withdraw_info.amount);
+						user_info.user_debt =
+							user_info.amount * pool_info.pool_acc / 10u32.pow(6).saturated_into();
+					}
+
+					StakingInfo::<T>::insert(&info_key, &user_info);
+
+					// update the lock balance of user
+					Self::update_lock(user_info.user.clone(), user_info.amount);
+				}
+			}
+		}
+
+		fn update_lock(who: T::AccountId, amount: BalanceOf<T>) {
+			<T as Config>::Currency::remove_lock(STAKING_ID, &who);
+			// reset lock
+			<T as Config>::Currency::set_lock(STAKING_ID, &who, amount, WithdrawReasons::all());
```

Comment:
```suggestion
			if amount == Zeor::zero() {
				<T as Config>::Currency::remove_lock(STAKING_ID, &who);
			} else {
				<T as Config>::Currency::set_lock(STAKING_ID, &who, amount, WithdrawReasons::all());
			}
```

Better to make them explicit